python/cudf/cudf/bindings/cudf_cpp.pyx

```diff
@@ -74,16 +76,53 @@ agg_ops = {
     'count_distinct': GDF_COUNT_DISTINCT,
 }
 
-def gdf_to_np_dtype(dtype):
-    """Util to convert gdf dtype to numpy dtype.
-    """
-    return np.dtype(gdf_dtypes[dtype])
+np_to_gdf_time_unit = {
+    's':  TIME_UNIT_s,
+    'ms': TIME_UNIT_ms,
+    'us': TIME_UNIT_us,
+    'ns': TIME_UNIT_ns,
+}
+
+gdf_to_np_time_unit = {
+    TIME_UNIT_s:  's',
+    TIME_UNIT_ms: 'ms',
+    TIME_UNIT_us: 'us',
+    TIME_UNIT_ns: 'ns',
+}
 
 def np_to_pa_dtype(dtype):
-    """Util to convert numpy dtype to PyArrow dtype
+    """Util to convert numpy dtype to PyArrow dtype.
     """
+    # special case when dtype is np.datetime64
+    if dtype.kind == 'M':
+        time_unit, _ = np.datetime_data(dtype)
+        if time_unit == 'ms':
+            return pa.date64()
+        if time_unit in np_to_gdf_time_unit:
+            # return a pa.Timestamp of the appropriate unit
+            return pa.timestamp(time_unit)
+        # return pa.date32() if time_unit is UNIX days,
+        # else pa.date64() since default is int64_t UNIX ms
+        return pa.date32() if time_unit == 'D' else pa.date64()
     return np_pa_dtypes[np.dtype(dtype).type]
```

Comment:
Any reason we can't replace this entire function with `pyarrow.from_numpy_dtype()`? Not sure if that existed when we wrote this 😄

Reply:
I think we could, but it looks like there's a mapping for `np.bool_` -> `pa.int8`. Stepping through in the debugger, it looks like `pa.py_buffer()` doesn't like when the ndarray returned by `devAry.copy_to_host()` has a boolean dtype.

Reply:
Yea, pyarrow booleans are a single bit per value as opposed to a byte per value. Could we just special case that one?

Reply:
By using the custom `np_to_pa_dtype` function and [this bool cast](https://github.com/rapidsai/cudf/blob/289fc3e29ba0c58fada3a3e998bd193bbca47edd/python/cudf/cudf/dataframe/numerical.py#L197), the current implementation is already special casing. From what I can tell, if we switch to using `from_numpy_dtype()`, the bool dtype will propagate down to the underlying Buffer (which is currently `np.int8`). That'd mean we'd have to cast the ndarray returned from `mem.copy_to_host()` on the CPU, which we are avoiding doing in our current implementation.